core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java

```diff
@@ -632,18 +649,52 @@ public CompletedTransactionsInfo() {
        * on a remote node. This might cause leaks, e.g. if the transaction is prepared, committed and prepared again.
        * Once marked as completed (because of commit or rollback) any further prepare received on that transaction are discarded.
        */
-      public void markTransactionCompleted(GlobalTransaction globalTx) {
-         log.tracef("Marking transaction %s as completed", globalTx);
-         completedTransactions.put(globalTx, timeService.time());
+      public void markTransactionCompleted(GlobalTransaction globalTx, boolean successful) {
+         if (trace) log.tracef("Marking transaction %s as completed", globalTx);
+         completedTransactions.put(globalTx, new CompletedTransactionInfo(timeService.time(), successful));
       }
 
       /**
-       * @see #markTransactionCompleted(org.infinispan.transaction.xa.GlobalTransaction)
+       * @see #markTransactionCompleted(GlobalTransaction, boolean)
        */
       public boolean isTransactionCompleted(GlobalTransaction gtx) {
-         if (completedTransactionsInfo == null)
+         if (completedTransactions.containsKey(gtx))
+            return true;
+
+         // Transaction ids are allocated in sequence, so any transaction with a smaller id must have already finished.
```

Comment:
this second part of the assumption is false. The ids are generated at the begin() so you can commit first transaction 10 and later transaction 1.